docs/specification.md

```diff
@@ -1,150 +1,354 @@
+required:
+  - names
+  - lengths
+inherent:
+  - lengths
+  - names
+  - sequences
+```
+
+This schema is the *seqcol schema*, and sequence collection objects in this structure are said to be the *canonical seqcol object representation*. Here's an example of a sequence collection organized into the canonical seqcol object representation:
+
+```json
+{
+  "lengths": [
+    248956422,
+    133797422,
+    135086622
+  ],
+  "names": [
+    "chr1",
+    "chr2",
+    "chr3"
+  ],
+  "sequences": [
+    "2648ae1bacce4ec4b6cf337dcae37816",
+    "907112d17fcb73bcab1ed1c72b97ce68",
+    "1511375dc2dd1b633af8cf439ae90cec"
+  ]
+}
+```
+
+This object would validate against the JSON-schema above. The object is a series of arrays with matching length (`3`), with the corresponding entries collated such that the first element of each array corresponds to the first element of each other array. For rationale for this structure over an array of annotated sequences, see *Footnote F1*. Implementations `MUST` provide at least the structure specified in this schema. Implementations `MAY` choose to extend this schema by adding additional attributes. This schema extends vanilla JSON-schema in two ways; first, it provides the `collated` qualifier. For further details about the rationale behind collated attributes, see *Footnote F2*. Second, it specifies the `inherent` qualifier. For further details about the rationale and examples of non-inherent attributes, see *Footnote F3*. Finally, another detail that may be unintuitive at first is that the `sequences` attribute is optional; for an explanation of why, see *Footnote F4*.
+
+##### Filter non-inherent attributes
+
+The `inherent` section in the seqcol schema is an extension of the basic JSON-schema format that adds specific functionality. Inherent attributes are those that contribute to the identifier; *non-inherent* attributes are not considered in computing the top-level digest. Attributes of a seqcol that are *not* listed as `inherent` `MUST NOT` contribute to the identifier; they are therefore excluded from the digest calculation. Therefore, if the canonical seqcol representation includes any non-inherent attributes, these must be removed before proceeding to step 2. In the simple example, there are no non-inherent attributes.
+
+#### Step 2: Apply RFC-8785 to canonicalize the value associated with each attribute individually.
+
+The [RFC-8785 JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785) (JCS) standardizes whitespace, character encodings, and other details that would cause inconsequential variations to yield different digests. For most use cases, the following Python fuction suffices:
+
+```python
+def canonical_str(item: dict) -> str:
+    """Convert a dict into a canonical string representation"""
+    return json.dumps(
+        item, separators=(",", ":"), ensure_ascii=False, allow_nan=False, sort_keys=True
+    )
```

Comment:
For most use cases it will suffice, though. maybe just making sure that's clear in the comments, and possibly even adding a warning to check for unicode?